field/TinyLang/Field/Typed/TypeChecker.hs

```diff
@@ -65,53 +59,53 @@ type MonadTypeError   m = MonadError TypeCheckError m
 type MonadTypeChecker m = ( MonadSupply m
                           , MonadScope m
                           , MonadTypeError m
+                          , MonadReader TyEnv m
                           )
 
+{-| == Type Environments
+-}
+
+-- TODO: Can be parameterised later
+type TyEnv = Set R.Var
+
+
 {-| @TypeChecker@ Transformer
 -}
-newtype TypeCheckerT e (m :: Type -> Type) a =
-    TypeChecker { runTypeCheckerT :: (ExceptT e (StateT Scope (SupplyT m))) a }
+newtype TypeCheckerT (m :: Type -> Type) a =
+    TypeChecker { runTypeCheckerT :: (ExceptT TypeCheckError (StateT Scope (ReaderT TyEnv (SupplyT m)))) a }
```

Comment:
Ah, indeed. I think the caller only needs mappings from external variable names to variable IDs. Let's keep it the way it's currently written, but I'll need to rethink it later.

Reply:
Ok, it makes sense. We only use `Scope` to ensure that a variable name always maps to the same variable, completely regardless of scoping. Scopes are resolved later in the renamer.

But then we now also resolve scopes in the type checker in addition to resolving them in the renamer. And we can't apply the renamer beforehand, because of loops.

Would the best way be to

1. unroll loops
2. rename (after moving the renamer to operate on untyped terms rather than typed ones)
3. type check

? Ok, I'll think about that.